src/lib/operators/join_hash/join_hash_steps.hpp

```diff
@@ -338,34 +338,77 @@ RadixContainer<T> partition_radix_parallel(const RadixContainer<T>& radix_contai
   radix_output.partition_offsets.resize(num_partitions);
   radix_output.null_value_bitvector = output_nulls;
 
-  // use histograms to calculate partition offsets
-  size_t offset = 0;
-  std::vector<std::vector<size_t>> output_offsets_by_chunk(chunk_offsets.size(), std::vector<size_t>(num_partitions));
-  for (size_t partition_id = 0; partition_id < num_partitions; ++partition_id) {
-    for (ChunkID chunk_id{0}; chunk_id < chunk_offsets.size(); ++chunk_id) {
-      output_offsets_by_chunk[chunk_id][partition_id] = offset;
-      offset += histograms[chunk_id][partition_id];
+  /**
+   * The following steps create the offsets that allow each concurrent job to write lock-less into the newly
+   * created RadixContainer. The input RadixContainer was a simple list of materialized values. The outcome
+   * container is consecutive vector, sorted by radix clusters (which are defined by the offsets).
+   * Input:
+   *  - `histograms` stores for each input chunk a histogram that counts the number of values per radix cluster.
+   *  - `chunk_offsets` stores the offsets -- denoting the number of elements of each chunk -- in the continuous
+   *     vector of materialized values.
+   *
+   * Simple example:
+   * - histograms for chunks 0 and 1 [4 0 3] & [5 2 7] (i.e., first radix cluster has 4 + 5 values)
+   * - first step: create offsets that denote write offsets per radix cluster and collect lengths
+   *   - result: offsets vectors are [0 0 0] [4 0 3] and lengths are [9 2 10]
+   * - second step: create prefix sum vector of [9 2 10] >> [9 11 10]
+   * - third step: adapt offset vectors to create the offsets that allow writing
+   *   all threads in parallel into a _consecutive_ vector
+   *   - result: [0 9 11] [4 9 14]
+   * - note: the third step would be superfluous, when each radix cluster is written to a separate vector
+   */
+  std::vector<size_t> output_offsets_by_chunk;
+  output_offsets_by_chunk.reserve(chunk_offsets.size() * num_partitions);
+
+  // Offset vector creation: first step
+  auto prefix_sums = std::vector<size_t>(num_partitions);  // stores the prefix sums
+  for (auto chunk_id = ChunkID{0}; chunk_id < chunk_offsets.size(); ++chunk_id) {
+    for (auto radix_cluster_id = size_t{0}; radix_cluster_id < num_partitions; ++radix_cluster_id) {
+      const auto radix_cluster_size = histograms[chunk_id][radix_cluster_id];
+      output_offsets_by_chunk.push_back(prefix_sums[radix_cluster_id]);
```

Comment:
Do we need push_back or can we use zero-initialization and indexes?

Reply:
Done.